nflow-engine/src/main/java/io/nflow/engine/internal/executor/WorkflowStateProcessor.java

```diff
@@ -267,15 +269,8 @@ private WorkflowInstance saveWorkflowInstanceState(StateExecutionImpl execution,
             ex);
         sleepIgnoreInterrupted(stateSaveRetryDelay);
       }
-    } while (internalRetryEnabled);
-    throw new IllegalStateException(format("Failed to save workflow instance %s new state", instance.id));
-  }
-
-  /**
-   * For unit testing only
-   */
-  void setInternalRetryEnabled(boolean internalRetryEnabled) {
-    this.internalRetryEnabled = internalRetryEnabled;
+    } while (!shutdownRequested.get());
```

Comment:
Check shutdown in catch block above. If shutting down, log error (failed, shutting down) and return the instance. If not, log error (failed, retrying) and go to sleep.
And check for shutdown in run method catch block also. If shutting down, log error (failed, shutting down) and stateProcessingFinished = true, else log error (failed, retrying) and got to sleep.